File: csv-query/python/src/csv_query/row.py
from __future__ import annotations

from .exceptions import UnknownColumnError


class Row:
    def __init__(self, values: dict[str, str]) -> None:
        self._values = dict(values)

    def get(self, column: str) -> str:
        if column not in self._values:
            raise UnknownColumnError(column)
        return self._values[column]

    @property
    def columns(self) -> list[str]:
        return list(self._values.keys())

    @property
    def values(self) -> dict[str, str]:
        return dict(self._values)

    def project(self, columns: list[str]) -> Row:
        projected = {}
        for col in columns:
            projected[col] = self.get(col)
        return Row(projected)

File: csv-query/typescript/src/Row.ts
import { UnknownColumnError } from './UnknownColumnError.js';

export class Row {
  private readonly values: ReadonlyMap<string, string>;

  constructor(values: Map<string, string> | Record<string, string>) {
    this.values = values instanceof Map
      ? new Map(values)
      : new Map(Object.entries(values));
  }

  get(column: string): string {
    const value = this.values.get(column);
    if (value === undefined) throw new UnknownColumnError(column);
    return value;
  }

  get columns(): string[] {
    return [...this.values.keys()];
  }

  project(columns: readonly string[]): Row {
    const projected = new Map<string, string>();
    for (const col of columns) {
      projected.set(col, this.get(col));
    }
    return new Row(projected);
  }

  toRecord(): Record<string, string> {
    const rec: Record<string, string> = {};
    for (const [k, v] of this.values) {
      rec[k] = v;
    }
    return rec;
  }
}
